### PR TITLE
Unbiased CSPRNG

### PR DESCRIPTION
**Pull request by paragonie-scott**
See #171

### CODE REVIEW

**Comment by paragonie-scott** [comment]
This adds a `randomInt()` function that should prove useful for other password options (e.g. diceware passphrase generation). It also should serve as an unbiased source of random integers in a given range if you need them elsewhere. :)

File: src/app.js [hunk 1]
```diff
@@ -287,15 +287,55 @@ var Encryptr = (function (window, console, undefined) {
...
+       * value to reduce the number of lookups
+       */
+      var mask = 1;
+      while (tmp > 0) {
+        if (bits % 8 === 0) {
+          bytes++;
+        }
+        bits++;
+        mask = mask << 1 | 1;
+        range >>= 1;
```
Comment (fx5) [inline src/app.js]: this should be `tmp` instead of `range`?

File: src/app.js [hunk 2]
```diff
@@ -287,15 +287,55 @@ var Encryptr = (function (window, console, undefined) {
...
+        result += charset[Encryptr.randomInt(0, max)];
       }
     }
     return result; // If you can't say something nice, don't say anything at all
   };
+  
+  Encryptr.prototype.randomInt = function(min, max) {
+    var i = rval = bits = bytes = 0;
+    var range = max - min;
+    if (range < 2) {
```
Comment (fx5) [inline src/app.js]: this should be `< 1`?

File: src/app.js [hunk 3]
```diff
@@ -287,15 +287,55 @@ var Encryptr = (function (window, console, undefined) {
...
+      }
+      
+      var values = new Uint32Array(bytes);
+      do {
+        window.crypto.getRandomValues(values);
+        
+        // Turn the random bytes into an integer
+        rval = 0;
+        for (i = 0; i < bytes; i++) {
+          rval |= (values[i] << (8 * i));
```
Comment (fx5) [inline src/app.js]: since values[i] are 32 bit values, the `|` adds a huge bias for numbers >= (2^8, 2^16...)

Comment (paragonie-scott) [inline src/app.js]: Sorry, I forgot to change it to a `Uint8Array`.

**Comment by chauffer** [comment]
:+1:

File: src/app.js (hunk 3 above)
Comment (paragonie-scott) [inline src/app.js]: (That `Uint8Array` change was made ~8 months ago, by the way.)

**Pull request closed without merging**